src/services/pcn-bridge/src/Bridge.cpp

```diff
@@ -1,143 +1,342 @@
+  if (tagged) {
+    Tins::Dot1Q &dot1 = ether.rfind_pdu<Tins::Dot1Q>();
+    bpdusize = dot1.payload_type() - LLC_HEADER_LEN;
+    start += DOT1Q_HEADER_LEN;
+  } else {
+    bpdusize = ether.payload_type() - LLC_HEADER_LEN;
+  }
+
+  // take last part of vector
+  std::vector<uint8_t>::const_iterator first = packet.begin() + start;
+  std::vector<uint8_t>::const_iterator last = packet.begin() + start + bpdusize;
+  std::vector<uint8_t> p(first, last);
+  getStp(vlan)->processBPDU(port_id, p);
+}
+
+void Bridge::broadcastPacket(Ports &port,
+                             polycube::service::PacketInMetadata &md,
+                             const std::vector<uint8_t> &packet) {
+  logger()->trace("Broadcasting packet..");
+
+  bool tagged = md.metadata[1];
+  uint16_t vlan = md.metadata[0];
+  logger()->trace("metadata vlan:{0} tagged:{1}", vlan, tagged);
+
+  std::lock_guard<std::mutex> guard(ports_mutex_);
+
+  for (auto &it : get_ports()) {
+    if (it->name() == port.name()) {
+      continue;
+    }
+    logger()->trace("Sending packet to port {0}", it->name());
+    it->send_packet_out(packet, tagged, vlan);
+  }
 }
 
-// Basic default implementation, place your extension here (if needed)
 std::shared_ptr<Ports> Bridge::getPorts(const std::string &name) {
-  // call default implementation in base class
   return BridgeBase::getPorts(name);
 }
 
-// Basic default implementation, place your extension here (if needed)
 std::vector<std::shared_ptr<Ports>> Bridge::getPortsList() {
-  // call default implementation in base class
   return BridgeBase::getPortsList();
 }
 
-// Basic default implementation, place your extension here (if needed)
 void Bridge::addPorts(const std::string &name, const PortsJsonObject &conf) {
+  std::lock_guard<std::mutex> guard(ports_mutex_);
+
   BridgeBase::addPorts(name, conf);
 }
 
-// Basic default implementation, place your extension here (if needed)
 void Bridge::addPortsList(const std::vector<PortsJsonObject> &conf) {
-  // call default implementation in base class
   BridgeBase::addPortsList(conf);
 }
 
-// Basic default implementation, place your extension here (if needed)
 void Bridge::replacePorts(const std::string &name,
                           const PortsJsonObject &conf) {
-  // call default implementation in base class
   BridgeBase::replacePorts(name, conf);
 }
 
-// Basic default implementation, place your extension here (if needed)
 void Bridge::delPorts(const std::string &name) {
-  // call default implementation in base class
+  std::lock_guard<std::mutex> guard(ports_mutex_);
+
   BridgeBase::delPorts(name);
 }
 
-// Basic default implementation, place your extension here (if needed)
 void Bridge::delPortsList() {
-  // call default implementation in base class
   BridgeBase::delPortsList();
 }
 
 std::shared_ptr<Fdb> Bridge::getFdb() {
-  throw std::runtime_error("Bridge::getFdb: Method not implemented");
+  if (fdb_ == nullptr || fdb_.use_count() == 0)
```

Comment:
I am curious, why do you have to check the use_count() ?

Reply:
I don't remember, it's useless actually :sweat_smile: